ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Iterator.hs

```diff
@@ -54,220 +51,206 @@ import           Ouroboros.Consensus.Storage.ImmutableDB.Impl.Index.Secondary
-  -> m (Either (WrongBoundError hash)
-               (Iterator hash m b))
-streamImpl dbEnv registry blockComponent mbStart mbEnd =
+  -> BlockComponent blk b
+  -> StreamFrom blk
+  -> StreamTo   blk
+  -> m (Either (WrongBoundError blk) (Iterator m blk b))
+streamImpl dbEnv registry blockComponent = \from to ->
     withOpenState dbEnv $ \hasFS OpenState{..} -> runExceptT $ do
-      lift $ either throwM return =<<
-        validateIteratorRange chunkInfo (forgetTipInfo <$> currentTip)
-          mbStart mbEnd
-
-      case currentTip of
-        Origin ->
-          -- If any of the two bounds were specified, 'validateIteratorRange'
-          -- would have thrown a 'ReadFutureSlotError'.
-          assert (isNothing mbStart && isNothing mbEnd) $ return mkEmptyIterator
-        NotOrigin tip -> do
-          WithHash endHash endChunkSlot <- fillInEndBound   currentIndex tip mbEnd
-          (secondaryOffset, start)      <- fillInStartBound currentIndex     mbStart
-
-          lift $ do
-            -- 'validateIteratorRange' will catch nearly all invalid ranges,
-            -- except for one: streaming from the regular block to the EBB in
-            -- the same slot. The EBB comes before the regular block, so these
-            -- bounds are invalid. However, to distinguish the EBB from the
-            -- regular block, as both have the same slot number, we need to
-            -- look at the hashes. 'validateIteratorRange' doesn't have enough
-            -- information to do that.
-            let WithHash _startHash startChunkSlot = start
-            when (startChunkSlot > endChunkSlot) $ do
-              let startSlot = chunkSlotToSlot chunkInfo startChunkSlot
-                  endSlot   = chunkSlotToSlot chunkInfo endChunkSlot
-              throwUserError $ InvalidIteratorRangeError startSlot endSlot
-
-            let ChunkSlot startChunk startRelSlot = startChunkSlot
-                startIsEBB   = relativeSlotIsEBB startRelSlot
-                curChunkInfo = CurrentChunkInfo currentChunk currentChunkOffset
-
-            -- TODO avoid rereading the indices of the start thunk. We read
-            -- from both the primary and secondary index in 'fillInStartBound'
-
-            iteratorState <- iteratorStateForChunk hasFS currentIndex registry
-              curChunkInfo endHash startChunk secondaryOffset startIsEBB
-
-            varIteratorState <- newTVarM $ IteratorStateOpen iteratorState
-
-            return $ mkIterator IteratorHandle
-              { itHasFS   = hasFS
-              , itIndex   = currentIndex
-              , itState   = varIteratorState
-              , itEnd     = endChunkSlot
-              , itEndHash = endHash
+      unless (validBounds from to) $
+        lift $ throwUserError $ InvalidIteratorRangeError from to
+
+      endChunkSlot                      <- checkEndBound   currentIndex currentTip to
+      (secondaryOffset, startChunkSlot) <- checkStartBound currentIndex            from
```

Comment:
Comment that this info corresponds to the lower bound, which might be inclusive. After opening the iterator, we advance it to the block after it (without reading the first one).